### PR TITLE
Add ReviewDB Plugin

### PR DESCRIPTION
**Pull request by mantikafasi**
very good plugin
Screenshot_53
Screenshot_52

### CODE REVIEW

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import { Toasts } from "../../../webpack/common";
+import { Review } from "../entities/Review";
+import { authorize, showToast } from "./Utils";
+
+
+const settings = Settings.plugins.ReviewDB;
+
+const API_URL = "https://manti.vendicated.dev";
+
+export const getReviews = async (discorid: number): Promise<Review[]> => {
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:11]: not a number

Comment (mantikafasi) [reply src/plugins/reviewDB/Utils/ReviewDBAPI.ts:11]: types arent real

**Comment by somerandomcloud** [comment]
i like shitpost pull requests, but can you please at least include some screenshots or something :sob:

**Comment by mantikafasi** [comment]
added ,do you love

**Comment by somerandomcloud** [comment]
horror

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts (hunk 1 above)
Comment (ArjixWasTaken) [reply src/plugins/reviewDB/Utils/ReviewDBAPI.ts]: NaN

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import { Review } from "../entities/Review";
+import { authorize, showToast } from "./Utils";
+
+
+const settings = Settings.plugins.ReviewDB;
+
+const API_URL = "https://manti.vendicated.dev";
+
+export const getReviews = async (discorid: string): Promise<Review[]> => {
+    const res = await fetch("https://manti.vendicated.dev/getUserReviews?snowflakeFormat=string&discordid=" + discorid.toString());
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:30]:
```suggestion
    const res = await fetch("https://manti.vendicated.dev/getUserReviews?snowflakeFormat=string&discordid=" + discorid);
```

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import { Toasts } from "../../../webpack/common";
+import { Review } from "../entities/Review";
+import { authorize, showToast } from "./Utils";
+
+
+const settings = Settings.plugins.ReviewDB;
+
+const API_URL = "https://manti.vendicated.dev";
+
+export const getReviews = async (discorid: string): Promise<Review[]> => {
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:29]:
```suggestion
export async function getReviews(discorid: string): Promise<Review[]> {
```

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    if (token === "") {
+        authorize();
+
+        Toasts.show({
+            message: "Please authorize to add a review.", options: { position: 1 },
+            id: "",
+            type: 0
+        });
+
+        return 2;
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:45]: what's 2

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+const API_URL = "https://manti.vendicated.dev";
+
+export const getReviews = async (discorid: string): Promise<Review[]> => {
+    const res = await fetch("https://manti.vendicated.dev/getUserReviews?snowflakeFormat=string&discordid=" + discorid.toString());
+    return await res.json() as Review[];
+};
+
+export const addReview = async (review: any): Promise<number> => {
+    const { token } = settings;
+    if (token === "") {
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:36]:
```suggestion
    if (!token) {
```

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            message: "Please authorize to add a review.", options: { position: 1 },
+            id: "",
+            type: 0
+        });
+
+        return 2;
+    }
+    review.token = token;
+
+    return await fetch(API_URL + "/addUserReview", { method: "POST", body: JSON.stringify(review) }).then(r => r.text()).then(
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:49]:
```suggestion
    return fetch(API_URL + "/addUserReview", { method: "POST", body: JSON.stringify(review) }).then(r => r.text()).then(
```

you should also set the proper content type header

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            let responseCode = 0;
+            // 0 means added ,1 means edited, 2 means error
+            if (res === "Added your review") {
+                responseCode = 0;
+            } else if (res === "Updated your review") {
+                responseCode = 1;
+            } else {
+                responseCode = 2;
+            }
+            return responseCode;
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:66]: inline this return and remove the variable

File: src/plugins/reviewDB/components/MessageButton.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { React } from "../../../webpack/common";
+const { PureComponent } = React;
+
+import { findByProps } from "../../../webpack";
+const { button, dangerous } = findByProps("button", "wrapper", "disabled");
+
+export default class MessageButton extends PureComponent<any> {
```
Comment (Vendicated) [inline src/plugins/reviewDB/components/MessageButton.tsx:25]: use a functional component

File: src/plugins/reviewDB/components/ReviewsView.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+const { eyebrow } = findByProps("eyebrow");
+const { bodyTitle } = findByProps("bodyTitle");
+const { section } = findByProps("section", "lastSection");
+const { defaultColor } = findByProps("defaultColor");
+
+interface IState {
+    reviews?: any[];
+}
+
+export const Queue = {
```
Comment (Vendicated) [inline src/plugins/reviewDB/components/ReviewsView.tsx:33]: Vencord already implements a Queue

File: src/plugins/reviewDB/components/ReviewsView.tsx [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    last: Promise.resolve(),
+    push(func: any) {
+        return (this.last = this.last.then(func));
+    }
+};
+
+export default class ReviewsView extends React.Component<any, IState> {
```
Comment (Vendicated) [inline src/plugins/reviewDB/components/ReviewsView.tsx:40]: use a functional component

File: src/plugins/reviewDB/entities/Review.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+export type Review = {
```
Comment (Vendicated) [inline src/plugins/reviewDB/entities/Review.ts:19]: use an interface

File: src/plugins/reviewDB/index.tsx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        this.ReviewsView = await import("./components/ReviewsView");
+        this.getLastReviewID = (await import("./Utils/ReviewDBAPI")).getLastReviewID;
+
+
+        console.log("ReviewDB Started");
+
+        const settings = Settings.plugins.ReviewDB;
+
+        settings.get = (key: string) => settings[key];
+        settings.set = (key: string, value: any) => settings[key] = value;
```
Comment (Vendicated) [inline src/plugins/reviewDB/index.tsx:59]: ?????

Consume directly

File: src/plugins/reviewDB/index.tsx [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+            }, description: "Authorize your account"
+        }
+    },
+
+    async start() {
+        this.ReviewsView = await import("./components/ReviewsView");
+        this.getLastReviewID = (await import("./Utils/ReviewDBAPI")).getLastReviewID;
+
+
+        console.log("ReviewDB Started");
```
Comment (Vendicated) [inline src/plugins/reviewDB/index.tsx:54]:
```suggestion
```

File: src/plugins/reviewDB/index.tsx [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+        settings.get = (key: string) => settings[key];
+        settings.set = (key: string, value: any) => settings[key] = value;
+
+
+        this.getLastReviewID(UserStore.getCurrentUser().id).then(lastreviewid => {
+            const storedLastReviewID: number = settings.get("lastreviewid", 0);
+            if (settings.get("notifyReviews", true) && storedLastReviewID < lastreviewid) {
+                if (storedLastReviewID !== 0) {
+                    showToast("You have new reviews on your profile");
```
Comment (Vendicated) [inline src/plugins/reviewDB/index.tsx:66]: use a timer to not bother the user before Discord has loaded

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts (hunk 4 above)
Comment (mantikafasi) [reply src/plugins/reviewDB/Utils/ReviewDBAPI.ts:45]: means user isnt authorized, in the promise.then() it doesnt clear the input box

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        }
+    );
+};
+
+export const deleteReview = async (reviewid: string): Promise<any> => {
+    const data: any = {
+        "token": settings.get("token", ""),
```
[the remark this answers is not in the document]
Comment (mantikafasi) [reply src/plugins/reviewDB/Utils/ReviewDBAPI.ts:73]: its a function I added,now I will remove

File: src/plugins/reviewDB/components/ReviewComponent.tsx [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+const buttonClassNames = findByProps("button", "wrapper", "disabled");
+const usernameClickable = findByProps("clickable", "username").clickable;
+const { defaultColor } = findByProps("defaultColor");
+
+
+interface IState {
+    profilePhoto: string,
+}
+
+export default class ReviewComponent extends React.Component<any, IState> {
```
[the remark this answers is not in the document]
Comment (mantikafasi) [reply src/plugins/reviewDB/components/ReviewComponent.tsx:47]: explode

File: src/plugins/reviewDB/components/ReviewComponent.tsx [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+const { message } = findByProps("message");
+const { groupStart } = findByProps("groupStart");
+const { wrapper } = findByProps("wrapper", "zalgo");
+const { cozy } = findByProps("cozy", "zalgo");
+const { contents } = findByProps("contents");
+const { getUserAvatarURL } = findByProps("getUserAvatarURL");
+const getUser = UserUtils.fetchUser;
+const buttonClassNames = findByProps("button", "wrapper", "disabled");
+const usernameClickable = findByProps("clickable", "username").clickable;
+const { defaultColor } = findByProps("defaultColor");
```
[the remark this answers is not in the document]
Comment (mantikafasi) [reply src/plugins/reviewDB/components/ReviewComponent.tsx:40]: WHAT AM I SUPPOSED TO DO

File: src/plugins/reviewDB/index.tsx (hunk 12 above)
Comment (mantikafasi) [reply src/plugins/reviewDB/index.tsx:59]: ok

**Comment by KraXen72** [comment]
(suggestions) maybe move the reviewDB view into a new tab in userProfile rather than userPopout? userPopout is already pretty cluttered with the latest updates.

**Comment by mantikafasi** [comment]
you need to scrool down to access it , adding another tab would be very bad ux

**Comment by mantikafasi** [comment]
8th November 2022

**Comment by somerandomcloud** [comment]
> 8th November 2022

weirdly ominous but ok

**Comment by mantikafasi** [comment]
-2 hours 38 minutes left

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+export async function addReview(review: any): Promise<number> {
+    const { token } = settings;
+    if (!token) {
+        authorize();
+
+        Toasts.show({
+            message: "Please authorize to add a review.", options: { position: 1 },
+            id: "",
+            type: 0
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:40]: don't hardcode this and pass a proper id (Toasts.generateId())

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+        .then(
+            res => {
+                Toasts.show({
+                    message: res, options: { position: 1 },
+                    id: "",
+                    type: 0
+                });
+
+                // 0 means added ,1 means edited, 2 means error
+                return (res === "Added your review") ? 0 : (res === "Updated your review") ? 1 : 2;
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:65]: return an enum value

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        );
+}
+
+export function deleteReview(reviewid: string): Promise<any> {
+    const data = {
+        "token": settings.token,
+        "reviewid": reviewid
+    };
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:74]: inline this

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+    return fetch(API_URL + "/deleteReview", { method: "POST", body: JSON.stringify(data) })
+        .then(r => r.json());
+}
+
+export function reportReview(reviewID: string) {
+    const data = {
+        "reviewid": reviewID,
+        "token": settings.get("token", "")
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:82]: inline this

File: src/plugins/reviewDB/Utils/ReviewDBAPI.ts [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+    fetch(API_URL + "/reportReview", {
+        method: "POST",
+        body: JSON.stringify(data)
+    })
+        .then(r => r.text())
+        .then(res => showToast(res));
+}
+
+export const getLastReviewID = async (userid: string): Promise<number> => {
+    console.log("Ur mom calling");
```
Comment (Vendicated) [inline src/plugins/reviewDB/Utils/ReviewDBAPI.ts:93]:
```suggestion
```

**Review by Vendicated — APPROVED** [review]

**Pull request merged**